crux_core/src/capability.rs

```diff
@@ -1,6 +1,119 @@
-/// TODO docs!
-pub trait Capability {}
+use std::sync::Arc;
 
-pub trait Capabilities<C> {
-    fn get(&self) -> &C;
+use futures::Future;
+
+use crate::{
+    channels::{Receiver, Sender},
+    Command,
+};
+
+// TODO docs!
+pub trait Capability<Ev> {
+    type MappedSelf<MappedEv>;
+
+    fn map_event<F, NewEvent>(&self, f: F) -> Self::MappedSelf<NewEvent>
+    where
+        F: Fn(NewEvent) -> Ev + Send + Sync + Copy + 'static,
+        Ev: 'static,
+        NewEvent: 'static;
+}
+
+pub trait CapabilitiesFactory<App, Ef>
+where
+    App: crate::App,
+{
+    fn build(context: CapabilityContext<Ef, App::Event>) -> App::Capabilities;
+}
+
+pub struct CapabilityContext<Ef, Event> {
+    inner: std::sync::Arc<ContextInner<Ef, Event>>,
+}
+
+struct ContextInner<Ef, Event> {
+    command_sender: Sender<Command<Ef>>,
+    event_sender: Sender<Event>,
+    spawner: crate::executor::Spawner,
+}
+
+impl<Ef, Ev> Clone for CapabilityContext<Ef, Ev> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+}
+
+impl<Ef, Ev> CapabilityContext<Ef, Ev>
+where
+    Ef: 'static,
+    Ev: 'static,
+{
+    pub(crate) fn new(
+        command_sender: Sender<Command<Ef>>,
+        event_sender: Sender<Ev>,
+        spawner: crate::executor::Spawner,
+    ) -> Self {
+        let inner = Arc::new(ContextInner {
+            command_sender,
+            event_sender,
+            spawner,
+        });
+
+        CapabilityContext { inner }
+    }
+
+    pub(crate) fn run_command(&self, cmd: Command<Ef>) {
+        self.inner.command_sender.send(cmd);
+    }
+
+    pub fn spawn(&self, f: impl Future<Output = ()> + 'static + Send) {
```

Comment:
> I can't quite put my finger on it, but I think something around this API should nod to the fact that this is essentially the capability's Shell interface, especially the effect method.

Yeah, broadly speaking I'm not against renaming/rejigging some of this stuff.  It's the most "user" (well capability implementor) visibile bit of this so yeah.

> It might be that the parts talking to the shell (effect) and the parts talking back to the app (send_event) should be separate?

I don't know.  I definitely see where you're coming from but wouldn't a capability be useless without both of them?  Like you'd just end up having two parameters/fields everywhere instead of one.

Exposing an API like `ctx.core.send_event()` could be nice mind you, although that would mean introducing more things so not sure it's actually a good idea...?

> Maybe that would also avoid the need to clone the context? Is the context doing to many jobs...?

I'm not sure reorganising things would remove the need for a clone.  Our futures are `'static` so they can't borrow.  They need to be able to send events to the core and since they can't borrow they need to own the thing that does that.  So there needs to be a clone _somewhere_.

We could maybe go down the path of associating a lifetime with our futures.  Not inconceivable that they could inherit a lifetime from the core.  But I do not know that it's worth it - don't think it's an expensive clone, and putting lifetimes on things could open up a whole new world of pain.  You'd still not be able to borrow the `&self` so would probably still need to do some `let ctx = &ctx` type thing. 

One "simplification" I was thinking about was making the futures `impl Future<Output=Ev>`, which would remove the need for `send_event`.  But then it'd be impossible to build any kind of subscription capability on top of that API, so not sure it's actually better?